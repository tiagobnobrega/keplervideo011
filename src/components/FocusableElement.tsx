import { isEqual } from 'lodash';
import React, { Ref, useState } from 'react';
import {
  StyleProp,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

interface FocusableElementProps extends TouchableOpacityProps {
  children?: React.ReactNode;
  onBlur?: () => void;
  onFocus?: () => void;
  getFocusState?: (value: boolean) => void;
  onFocusOverrideStyle: StyleProp<ViewStyle>;
  onPress?: () => void;
  hasTVPreferredFocus?: boolean;
  focusableElementRef?: Ref<TouchableOpacity>;
  style: StyleProp<ViewStyle>;
}

const FocusableElement = ({
  focusableElementRef,
  children,
  onPress,
  onBlur,
  onFocus,
  getFocusState,
  onFocusOverrideStyle,
  style,
  hasTVPreferredFocus,
  ...otherProps
}: FocusableElementProps) => {
  const [isFocused, setIsFocused] = useState(false);
  const focusHandler = () => {
    setIsFocused(true);
    getFocusState?.(true);
    onFocus?.();
  };
  const blurHandler = () => {
    setIsFocused(false);
    getFocusState?.(false);
    onBlur?.();
  };

  return (
    <TouchableOpacity
      ref={focusableElementRef}
      activeOpacity={1}
      hasTVPreferredFocus={hasTVPreferredFocus}
      onFocus={focusHandler}
      onBlur={blurHandler}
      onPress={onPress}
      style={[style, isFocused ? onFocusOverrideStyle : undefined]}
      testID={otherProps.testID}
      {...otherProps}>
      {children}
    </TouchableOpacity>
  );
};

const focusableElementUnchanged = (
  prevProps: FocusableElementProps,
  nextProps: FocusableElementProps,
) => {
  return isEqual(prevProps, nextProps);
};

export default React.memo(FocusableElement, focusableElementUnchanged);
