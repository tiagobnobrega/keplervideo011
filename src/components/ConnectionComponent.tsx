import { Typography } from '@amzn/kepler-ui-components';
import {
  NetInfoConnectedDetails,
  NetInfoState,
  refresh,
} from '@amzn/keplerscript-netmgr-lib';
import MaterialIcons from '@amzn/react-native-vector-icons/MaterialIcons';
import { useFocusEffect } from '@amzn/react-navigation__core';
import React, { useCallback, useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { NetworkStatus } from '../constants';
import { COLORS } from '../styles/Colors';
import { scaleUxToDp } from '../utils/pixelUtils';
import FocusableElement from './FocusableElement';

interface ConnectionComponentProps {
  testID: string;
}
const ConnectionComponent = ({ testID }: ConnectionComponentProps) => {
  const [wifiStatus, setWifiStatus] = useState<string>('');
  const [wifiDetails, setWifiDetails] =
    useState<NetInfoConnectedDetails | null>(null);

  const onSuccess = (state: NetInfoState) => {
    const wifiStatusCheck = state.isConnected && state.isInternetReachable;
    const wifiStatusText = wifiStatusCheck
      ? NetworkStatus.CONNECTED
      : NetworkStatus.NOT_CONNECTED;
    setWifiStatus(wifiStatusText);
    if (state.isWifiEnabled && wifiStatusCheck) {
      setWifiDetails(state.details);
    } else {
      setWifiDetails(null);
    }
  };

  const onError = () => {
    setWifiStatus(NetworkStatus.NOT_CONNECTED);
    setWifiDetails(null);
  };

  const handleRefresh = () => {
    refresh()
      .then(state => {
        onSuccess(state);
      })
      .catch(() => {
        onError();
      });
  };

  useFocusEffect(
    useCallback(() => {
      handleRefresh();
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []),
  );

  return (
    <View style={styles.container} testID={testID}>
      <View style={styles.detailContainer}>
        <FocusableElement
          style={styles.refreshIcon}
          onFocusOverrideStyle={styles.refreshButtonFocus}
          onPress={handleRefresh}
          testID="refresh-icon">
          <MaterialIcons name={'refresh'} size={40} color={COLORS.WHITE} />
        </FocusableElement>
        <View>
          <Typography
            variant={'title'}
            numberOfLines={1}
            color={COLORS.WHITE}
            testID="internet-status"
            style={styles.detailText}>
            Internet Status: {wifiStatus}
          </Typography>

          {wifiDetails && Object.keys(wifiDetails).length ? (
            <Typography
              variant={'title'}
              testID="wifi-name"
              numberOfLines={1}
              color={COLORS.WHITE}
              style={styles.detailText}>
              {/* @ts-ignore */}
              Wifi Name: {wifiDetails?.ssid}
            </Typography>
          ) : null}
        </View>
      </View>
    </View>
  );
};

export default ConnectionComponent;

export const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: scaleUxToDp(20),
    right: 0,
    margin: scaleUxToDp(20),
    marginRight: scaleUxToDp(60),
    alignItems: 'center',
  },
  detailContainer: {
    flexDirection: 'row',
    alignContent: 'center',
  },
  detailText: {
    color: COLORS.WHITE,
    fontSize: scaleUxToDp(20),
    margin: scaleUxToDp(10),
    marginVertical: scaleUxToDp(2),
  },
  refreshIcon: {
    marginRight: scaleUxToDp(10),
    marginTop: scaleUxToDp(12),
    width: scaleUxToDp(50),
    height: scaleUxToDp(50),
    alignItems: 'center',
    justifyContent: 'center',
  },
  refreshButtonFocus: {
    backgroundColor: `${COLORS.SMOKE_WHITE}50`,
    borderRadius: scaleUxToDp(50),
  },
});
