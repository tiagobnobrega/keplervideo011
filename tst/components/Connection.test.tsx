import { useFocusEffect } from '@amzn/react-navigation__core';
import '@testing-library/jest-native/extend-expect';
import {
  fireEvent,
  render,
  RenderAPI,
  waitFor,
} from '@testing-library/react-native';
import React from 'react';
import ConnectionComponent from '../../src/components/ConnectionComponent';
import FocusableElement from '../../src/components/FocusableElement';
global.fetch = jest.fn();

jest.mock('@amzn/keplerscript-netmgr-lib', () => ({
  refresh: jest.fn(),
}));

jest.mock('@amzn/react-navigation__core', () => {
  return {
    useFocusEffect: jest.fn(),
    NavigationContainer: ({ children }: { children: React.ReactNode }) => (
      <>{children}</>
    ),
  };
});

jest.mock('@amzn/kepler-ui-components', () => ({
  __esModule: true,
  Typography: jest.fn(),
}));

jest.mock('@amzn/keplerscript-netmgr-lib', () => ({
  __esModule: true,

  fetch: jest.fn(),
  NetInfoConnectedDetails: jest.fn(),
  NetInfoState: jest.fn(),
  refresh: jest.fn().mockResolvedValue({}),
}));

const handleRefreshError = (): Promise<void> => {
  throw new Error('I should fail');
};

describe('Connection Component Test Cases', () => {
  let tree: RenderAPI;
  beforeEach(() => {
    tree = render(<ConnectionComponent testID="connection-component" />);
  });

  it('renders correctly with styles', () => {
    const snap = tree;
    expect(snap).toMatchSnapshot();
  });

  it('renders correctly with initial state', () => {
    const { getByTestId } = tree;
    expect(getByTestId('connection-component')).toBeTruthy();
  });

  it('does not display wifi details when wifiDetails is empty object', () => {
    const { queryByText } = tree;
    expect(queryByText('Wifi Name:')).toBeNull();
  });

  it('styles applied correctly', () => {
    const { getByTestId } = tree;
    const container = getByTestId('connection-component');
    expect(container).toHaveStyle({ position: 'absolute' });
  });

  it('returns the number of keys when wifiDetails has keys', () => {
    tree;
    const wifiDetails = {
      ssid: 'TestWifi',
      password: 'password123',
      security: 'WPA2',
    };
    expect(Object.keys(wifiDetails).length).toBe(3);
  });

  it('FocusableElement responds to focus and press events', () => {
    // Mock onPress function
    const onPressMock = jest.fn();
    // Render the dummy FocusableElement with a testID and required params
    const { getByTestId } = render(
      <FocusableElement
        onPress={onPressMock}
        testID="focusable-element"
        onFocusOverrideStyle={undefined}
        style={undefined}
      />,
    );
    fireEvent.press(getByTestId('focusable-element'));
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it('FocusableElement handles focus and blur events correctly', () => {
    // Mock onFocus and onBlur callback functions
    const onFocusMock = jest.fn();
    const onBlurMock = jest.fn();

    // Render the FocusableElement with onFocus and onBlur props
    const { getByTestId } = render(
      <FocusableElement
        onFocus={onFocusMock}
        onBlur={onBlurMock}
        testID="focusable-element"
        onFocusOverrideStyle={undefined}
        style={undefined}
      />,
    );
    fireEvent(getByTestId('focusable-element'), 'focus');
    expect(onFocusMock).toHaveBeenCalled();
    fireEvent(getByTestId('focusable-element'), 'blur');
    expect(onBlurMock).toHaveBeenCalled();
  });

  it('calls handleRefresh when screen is in focus', async () => {
    const handleRefresh = jest.fn();
    render(<ConnectionComponent testID="connection-component" />);
    await (useFocusEffect as jest.Mock).mock.calls[0][0]();
    handleRefresh();
    waitFor(() => {
      expect(handleRefresh).toHaveBeenCalled();
    });
  });

  it('onError is called when refresh throws an error', () => {
    expect.assertions(1);
    const mockOnError = jest.fn();
    render(<ConnectionComponent testID="connection-component" />);
    try {
      handleRefreshError();
    } catch (error) {
      mockOnError();
    }
    expect(mockOnError).toHaveBeenCalled();
  });
});
