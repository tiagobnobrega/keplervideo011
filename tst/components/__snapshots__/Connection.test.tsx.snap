// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection Component Test Cases renders correctly with styles 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "margin": 20,
      "marginRight": 60,
      "position": "absolute",
      "right": 0,
      "top": 20,
    }
  }
  testID="connection-component"
>
  <View
    style={
      Object {
        "alignContent": "center",
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        Object {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "height": 50,
          "justifyContent": "center",
          "marginRight": 10,
          "marginTop": 12,
          "opacity": 1,
          "width": 50,
        }
      }
      testID="refresh-icon"
    />
    <View />
  </View>
</View>
`;
